src/core/pio-cart.h

```diff
@@ -0,0 +1,149 @@
+
+namespace PCSX {
+
+class PIOCart {
+  public:
+    PIOCart() : m_pal(this) {
+        m_detachedMemory = static_cast<uint8_t *>(calloc(64 * 1024, 1));
+        if (m_detachedMemory == NULL) {
+            g_system->message("%s", _("Error allocating memory!"));
+        } else {
+            memset(m_detachedMemory, 0xff, static_cast<size_t>(64 * 1024));
+        }
+    }
+
+    void setLuts();
+
+    bool getSwitch() { return m_switchOn; }
+    void setSwitch(bool on) { m_switchOn = on & 1; }
+
+    uint8_t read8(uint32_t address);
+    uint16_t read16(uint32_t address) { return 0xffff; }      // 2 reads?
+    uint32_t read32(uint32_t address) { return 0xffffffff; }  // 4 read?
+    void write8(uint32_t address, uint8_t value);
+    void write16(uint32_t address, uint16_t value) {}  // 2 writes?
+    void write32(uint32_t address, uint32_t value) {}  // 4 writes?
+
+  private:
+    bool m_switchOn = true;
+    uint8_t *m_detachedMemory = NULL;
+
+    class PAL {
+      public:
+        PAL(PIOCart *parent) : m_flashMemory(this), m_pio(parent) {}
+
+        uint8_t read8(uint32_t address);
+        void write8(uint32_t address, uint8_t value);
+
+        void reset() {
+            m_flashMemory.resetFlash();
+            m_bank = 0;
+        }
+        void setLUTFlashBank(uint8_t bank);
+
+        friend class PIOCart;
+
+      private:
+        class FlashMemory {
+          public:
+            FlashMemory(PAL *parent) : m_pal(parent) {
+                m_softwareID = static_cast<uint8_t *>(calloc(64 * 1024, 1));
+                if (m_softwareID == NULL) {
+                    g_system->message("%s", _("Error allocating memory!"));
+                } else {
+                    for (int i = 0; i < (64 * 1024) - 1; i += 2) {
+                        m_softwareID[i] = 0xbf;
+                        m_softwareID[i + 1] = 0x10;
+                    }
+                }
+            }
+
+            void write8(uint32_t address, uint8_t value);
+
+            void softwareDataProtectEnablePageWrite() {
+                m_dataProtectEnabled = true;
+                m_pageWriteEnabled = true;
+                // To-do: Grab address/page from next write?
+            }
+
+            void softwareDataProtectDisable() { m_dataProtectEnabled = false; }
+            void softwareChipErase() {}
+            void enterSoftwareIDMode() {
+                setLUTSoftwareID();
+                resetCommandBuffer();
+            }
+
+            void exitSoftwareIDMode() {
+                setLUTNormal();
+                resetCommandBuffer();
+            }
+
+            bool checkCommand();
+
+            void resetCommandBuffer() {
+                memset(m_commandBuffer, 0, 6);
+                m_busCycle = 0;
+            }
+
+            void resetFlash() {
+                resetCommandBuffer();
+                m_dataProtectEnabled = true;
+                m_pageWriteEnabled = false;
+            }
+
+            void sectorErase() {}
```

Comment:
We might need the address of the sector? And it's going to memset it to 0xff.

Reply:
Looks like this was from one of the earlier flash chips I investigated. I did go ahead and implemented the chip-erase command